nameresolution/mdns/mdns.go

```diff
@@ -258,58 +321,168 @@ func (m *resolver) ResolveID(req nameresolution.ResolveRequest) (string, error)
+
+		// once will only be set for the first browser.
+		once = new(sync.Once)
+	})
+
+	// there is a window of time where we may not have subscribed
+	// to the appID yet but the first browser has already published
+	// the address and updated the cache. We should recheck the cache
+	// now that we are subscribed to ensure we get the address regardless.
+	// There should be no possibility that the first browser will update
+	// the cache before it's own execution context has reach this check so
+	// it can't short-circuit the subscription.
+	if addr := m.nextIPv4Address(req.ID); addr != nil {
+		return *addr, nil
+	}
+	if addr := m.nextIPv6Address(req.ID); addr != nil {
+		return *addr, nil
 	}
 
-	return addr, err
+	select {
+	case addr := <-sub.AddrChan:
+		// only 1 subscriber should have set the once var so
+		// this block should only get invoked once too.
+		if once != nil {
+			once.Do(func() {
+				// trigger the background refresh for additional addresses.
+				m.refreshChan <- req.ID
+
+				// block on the done channel as this signals that we have
+				// published the address to all other subscribers first.
+				<-done
+			})
+		}
+		return addr, nil
+	case err := <-sub.ErrChan:
+		if once != nil {
+			once.Do(func() {
+				// block on the done channel as this signals that we have
+				// published the error to all other subscribers first.
+				<-done
+			})
+		}
+		return "", err
+	}
 }
 
-// browseFirstOnly will perform a mDNS network browse for an address
+// browseOne will perform a mDNS network browse for an address
 // matching the provided app id. It will return the first address it
 // receives and stop browsing for any more.
-func (m *resolver) browseFirstOnly(ctx context.Context, appID string) (string, error) {
-	var addr string
+// This must be called in a sync.Once block to avoid concurrency issues.
+func (m *resolver) browseOne(ctx context.Context, appID string, done chan bool) {
+	go func() {
+		var addr string
+
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		// onFirst will be invoked on the first address received.
+		// Due to the asynchronous nature of cancel() there
+		// is no guarantee that this will ONLY be invoked on the
+		// first address. Ensure that multiple invocations of this
+		// function are safe.
+		onFirst := func(ip string) {
+			addr = ip
+			cancel() // cancel to stop browsing.
+		}
 
-	ctx, cancel := context.WithTimeout(ctx, firstOnlyTimeout)
-	defer cancel()
+		m.logger.Debugf("Browsing for first mDNS address for app id %s", appID)
 
-	// onFirst will be invoked on the first address received.
-	// Due to the asynchronous nature of cancel() there
-	// is no guarantee that this will ONLY be invoked on the
-	// first address. Ensure that multiple invocations of this
-	// function are safe.
-	onFirst := func(ip string) {
-		addr = ip
-		cancel() // cancel to stop browsing.
-	}
+		err := m.browse(ctx, appID, onFirst)
+		if err != nil {
+			m.pubErrToSubs(appID, err)
```

Comment:
@jjcollinge I think true should be set on the `done` channel after this line.

Reply:
Yes good catch thanks.

Reply:
Yes good catch, thanks.